### ISSUE TITLE
Bug, constructor does not check return type

### ISSUE THREAD

**Issue by gryumov** [JuliaLang/julia#47209]
Constructor does not check return type

```
julia> struct A
           x::Int64

           function A(f::Int64, h::Int64)::Int64 # <---
               return new(f)
           end
       end

julia> A(10, 10)
A(10)
```

```
julia> versioninfo()
Julia Version 1.8.2
Commit 36034abf26 (2022-09-29 15:21 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin21.6.0)
  CPU: 8 × Apple M1 Pro
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, apple-m1)
  Threads: 8 on 6 virtual cores
Environment:
  JULIA_EDITOR = atom  -a
  JULIA_NUM_THREADS = 8
```

**Comment by vtjnash** [comment]
fixed on master, possible by #44664

```
julia> A(10, 10)
ERROR: MethodError: Cannot `convert` an object of type A to an object of type Int64

Closest candidates are:
  convert(::Type{T}, ::Ptr) where T<:Integer
   @ Base pointer.jl:23
  convert(::Type{T}, ::T) where T<:Number
   @ Base number.jl:6
  convert(::Type{T}, ::Number) where T<:Number
   @ Base number.jl:7
  ...

Stacktrace:
 [1] A(f::Int64, h::Int64)
   @ Main ./REPL[1]:5
 [2] top-level scope
   @ REPL[2]:1
```

**Comment by jakobnissen** [comment]
Would it perhaps be a good idea to add a test case for this?
